REGRESSION (295259@main): thread-safety assertion crashes in some imported/w3c/web-platform-tests/html/canvas/offscreen/manual/text/canvas.2d.offscreen.worker.*.html tests

darinadler opened the pull request:
#### ac0859625ba1a79c85e4d9810eae97a058e26b05
<pre>
REGRESSION (295259@main): thread-safety assertion crashes in some imported/w3c/web-platform-tests/html/canvas/offscreen/manual/text/canvas.2d.offscreen.worker.*.html tests
<a href="https://rdar.apple.com/153002399">rdar://153002399</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=294285">https://bugs.webkit.org/show_bug.cgi?id=294285</a>

Reviewed by Matt Woodrow.

Revert code changes for <a href="https://bugs.webkit.org/show_bug.cgi?id=290042">https://bugs.webkit.org/show_bug.cgi?id=290042</a> from
295259@main because they used ImageBuffer in a way that&apos;s not thread safe.
Kept the tests that were added, expect a failure for now.

* LayoutTests/TestExpectations: Expect failure for canvas.2d.fillText-FontFace.html.

* LayoutTests/platform/ios/TestExpectations: Remove expected crashes.
* LayoutTests/platform/mac-wk2/TestExpectations: Ditto.

* Source/WebCore/html/OffscreenCanvas.cpp:
(WebCore::OffscreenCanvas::commitToPlaceholderCanvas):
* Source/WebCore/html/canvas/PlaceholderRenderingContext.cpp:
(WebCore::PlaceholderRenderingContextSource::setPlaceholderBuffer):
(WebCore::PlaceholderRenderingContextSource::setContentsToLayer):
* Source/WebCore/html/canvas/PlaceholderRenderingContext.h:
Revert changes.

Canonical link: <a href="https://commits.webkit.org/296494@main">https://commits.webkit.org/296494@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7bfb403e6fd38a3bad441e28571c46c2e8ccf7f4

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/108723 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/28384 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/18808 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/113934 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/59107 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/110686 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/29073 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/36946 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/82598 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/111671 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/23088 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/97933 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/63035 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/22508 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/16069 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/58634 "Built successfully") | 
| [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/101271 "Build is in progress. Recent messages:") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/92458 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/16119 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/117055 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/35777 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/26403 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/91615 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/36150 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/94200 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/91421 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/23281 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/36322 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/14082 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/31647 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/35677 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/41211 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/35387 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/38734 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/37065 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

mattwoodrow commented on Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBackendProxy.h:
```diff
@@ -77,15 +77,17 @@ class RemoteImageBufferProxyFlushState;
 
 class RemoteRenderingBackendProxy
-    : public IPC::Connection::Client, public RefCounted<RemoteRenderingBackendProxy>, SerialFunctionDispatcher {
+    : public IPC::Connection::Client, public ThreadSafeRefCountedAndCanMakeThreadSafeWeakPtr<RemoteRenderingBackendProxy>, private SerialFunctionDispatcher {
```
I don't think we should do this, `RemoteRenderingBackendProxy` isn't really thread safe, and allowing it to be destroyed on other threads will likely cause new issues.

Instead I think we should make `ImageBuffer` not-thread safe refcounted (since it is also not safe to use those from different threads), which would have hopefully caught this sooner. Possibly there are cases where this current state is load-bearing though :(

I think ultimately the regressing PR was incorrect (bad reviewer!), and we should not be storing a ref to an `ImageBuffer` that gets accessed from multiple threads.

mattwoodrow replied:
I think the right fix is to use the existing cross-thread serialization of `ImageBuffer` that is done by `PlaceholderRenderingContextSource::setPlaceholderBuffer`, and retain a ref at that point.

Quite a few possible races though between `setPlaceholderBuffer`, the main-thread lambda and `setContentsToLayer`.

I'd be happy to take over this if you'd prefer.

darinadler replied:
I’ll be happy to take another crack at this with your suggestion. But if it’s still not done and you’re ready to start working on this, feel free to grab it; I’ll watch bugs.webkit.org and you should reassign it to yourself.

darinadler replied:
It seems like the first step should be to revert the incorrect fix from https://github.com/WebKit/WebKit/commit/153da00a252619799ba4b32cd0ac6c5b8faf6a35 and then the next step would be to make `ImageBuffer` not thread-safe reference counted and fix all the issues with that. Then we can go back and try to fix [bug 290042](https://bugs.webkit.org/show_bug.cgi?id=290042) another way.

mattwoodrow replied:
My concern is that removing threadsafe refcounting from `ImageBuffer` might be non-trivial, and reverting https://github.com/WebKit/WebKit/commit/153da00a252619799ba4b32cd0ac6c5b8faf6a35 reinstates a known rendering issue. 

Building on top of the current main seems like the shortest path forward to a known-good state, and then we can hopefully clean up the refcounting as a follow-up.

Not a strongly held opinion though, happy to go either way.

darinadler replied:
I’m going to reinstate the known rendering issue. You can redo the fix for [bug 290042](https://bugs.webkit.org/show_bug.cgi?id=290042) in whatever way you like, as quickly as you like. There’s no good reason to leave a known thread-safety crash in the tree.

I don’t agree that we need to "build on top of the current main"; you can easily reapply the fix and make a correct version.

darinadler committed:
REGRESSION (295259@main): thread-safety assertion crashes in some imported/w3c/web-platform-tests/html/canvas/offscreen/manual/text/canvas.2d.offscreen.worker.*.html tests
rdar://153002399
https://bugs.webkit.org/show_bug.cgi?id=294285

Reviewed by Matt Woodrow.

Revert code changes for https://bugs.webkit.org/show_bug.cgi?id=290042 from
295259@main because they used ImageBuffer in a way that's not thread safe.
Kept the tests that were added, expect a failure for now.

* LayoutTests/TestExpectations: Expect failure for canvas.2d.fillText-FontFace.html.

* LayoutTests/platform/ios/TestExpectations: Remove expected crashes.
* LayoutTests/platform/mac-wk2/TestExpectations: Ditto.

* Source/WebCore/html/OffscreenCanvas.cpp:
(WebCore::OffscreenCanvas::commitToPlaceholderCanvas):
* Source/WebCore/html/canvas/PlaceholderRenderingContext.cpp:
(WebCore::PlaceholderRenderingContextSource::setPlaceholderBuffer):
(WebCore::PlaceholderRenderingContextSource::setContentsToLayer):
* Source/WebCore/html/canvas/PlaceholderRenderingContext.h:
Revert changes.

Canonical link: https://commits.webkit.org/296494@main
